### ISSUE TITLE
Shutil same file error

### ISSUE THREAD

**Issue by miraodasilva** [wandb/client#1593]
**Describe the bug**
Shutil same file error

**To Reproduce**
Log input audio at beginning of every epoch with 
```
wandb.log({"Input audio": [wandb.Audio(inputs[0].cpu().squeeze().numpy(), sample_rate=train_dataset.get_audio_rate())]})
```
**Expected behavior**
Logged without issues.

**Screenshots**
![image](https://user-images.githubusercontent.com/17463598/101655999-20ef4d00-3a3a-11eb-808c-15f7ccf5ba75.png)

**Operating System**
 - OS: Ubuntu 20.04
 - Browser chrome
 - Wandb Version 0.10.11

**Additional context**
Audio is displayed properly on wandb. However, code crashes with this error after 2 epochs. Replicated multiple times on different machines. Handling the exception during the log does not seem to work for some reason.

**Comment by vanpelt** [comment]
Hey @miraodasilva we'll need to get a fix in for the underlying issue, but until then you can fix this by not logging the same audio multiple times.  You could log these inputs once in their own media panel and then if you want to associate predictions or other media to them you can log a `wandb.Table` with different columns at each step.

**Comment by tssweeney** [comment]
Hey @miraodasilva we are digging into this now. Two questions:

1. Do you mind logging your audio outside of the array? (ie. `wandb.log({"Input Audio": wandb.Audio(...)})`)
2. Do you mind sharing some context around this call? The error seems to imply that the same audio file is being logged to the same key for the same training step. Are you perhaps making this same call a few lines prior with `commit=False`?

**Comment by miraodasilva** [comment]
Hi! Sure, no problem. However I am only logging once per epoch and I am shuffling the dataset so it should not be the same file. I'm not sure what's going on honestly. I will get back to you if I have any more info.

**Comment by miraodasilva** [comment]
> Hey @miraodasilva we are digging into this now. Two questions:
> 
> 1. Do you mind logging your audio outside of the array? (ie. `wandb.log({"Input Audio": wandb.Audio(...)})`)
> 2. Do you mind sharing some context around this call? The error seems to imply that the same audio file is being logged to the same key for the same training step. Are you perhaps making this same call a few lines prior with `commit=False`?

Actually, I am investigating this as we speak. The last training step should have commit=False to allow for further validation logs within the same step. If I am accidentally doing this in the penultimate step rather than the last, that could certainly be an issue. The concrete line was
```
wandb.log(logs, commit=True if i < (len(loader) - 1) else False)
```

**Comment by tssweeney** [comment]
This seems like a possible lead (I am not sure how your loop is constructed, but the condition shared may be suspect). By chance, does your `logs` object contain an `Input Audio` key?

**Comment by miraodasilva** [comment]
It does, yes. The concrete line is
```
logs.update({"Input audio": [wandb.Audio(inputs[0].cpu().squeeze().numpy(), sample_rate=train_dataset.get_audio_rate())]})
```

**Comment by tssweeney** [comment]
Ok, just so we are on the same page, this is how I understand the logic to be:
```{python}
for i in range(len(loader)):
   logs = {}
   # ... other code
   logs.update({"Input audio": [wandb.Audio(...)]})
   # ... other code
   wandb.log(logs, commit=True if i < (len(loader) - 1) else False)

# then some additional logging?
```

If you don't mind sharing explicitly the key code blocks of interest, I think we can get to the bottom of this.

**Comment by miraodasilva** [comment]
I think a rather large problem is the fact that I was accidentally initializing the logs to an empty dict outside the training loop, so logs were accumulated for the entire loop.

**Comment by miraodasilva** [comment]
So basically the logs = {} was above the first line, rather than on the second line, as you show.

**Comment by miraodasilva** [comment]
That would mean I am accumulating a huge log throughout the entire epoch. However it would contain only one audio given that I add the audio to the log only on iteration 0.

**Comment by tssweeney** [comment]
Do you mind sharing your code snippet? I have a few hypotheses and it might be easier to discuss your concrete implementation.

**Comment by miraodasilva** [comment]
Unfortunately I can't, as this is a project still in development and that would imply releasing the main training loop.

**Comment by miraodasilva** [comment]
One thing I should mention, below I am Saving a video with the same audio:
```
# Visualize inputs
                    logs.update(
                        {
                            "Input audio": [
                                wandb.Audio(
                                    inputs[0].cpu().squeeze().numpy(), sample_rate=train_dataset.get_audio_rate()
                                )
                            ]
                        }
                    )
                    print("logged audio.")
                    sample_path = save_av_sample(
                        vinputs[0],
                        train_dataset.get_video_rate(),
                        audio=inputs[0],
                        audio_rate=train_dataset.get_audio_rate(),
                    )
                    logs.update({"Input video": wandb.Video(sample_path, fps=25, format="mp4")})
```

**Comment by vanpelt** [comment]
Even some basic pseudo code with random data that reproduces this issue would be really helpful.

**Comment by miraodasilva** [comment]
Ok here is a tear down of what's going on. The loop is something along these lines:
```
for e in range(epochs):
    # Source of the insanity is the line below. It should be inside the inner loop, not this one
    logs = {}
    for i in range(len(loader)):
        # ... other code
        logs.update({"Input audio": [wandb.Audio(...)],"Input video": [wandb.Video(...)]})
        # ... other code
        wandb.log(logs, commit=True if i < (len(loader) - 1) else False)
    
    #commit is true here, as default, so epoch AND last iter metrics are logged.
    wand.log("Epoch": e)
```
**Consequence:**
Same audio and video are logged every single iteration. By printing the log items every time this is clear
![image](https://user-images.githubusercontent.com/17463598/101670637-52bcdf80-3a4b-11eb-890b-eeff9550d844.png)
So it is logging the same video and audio over and over and over again. However, each one is recorded in the files on wandb. The level of insanity became clear when I checked the files and saw that there were 10k+ audio files under media/audio. I'm not sure if the same is happening with video, but the video folder does not appear (maybe because too many files cant be displayed). 

**Source of error**
The run crashes arbitrarily after around 1 or 2 epochs. Clearly it doesn't mind posting the exact same file over and over again. I would suggest that the bug has to do with logging too many audio files, to a point where the name/id is repeated. Frankly I'm not sure. Sorry about the lack of initiative but I can't really afford (in terms of time) to devote much more attention to the issue. By moving log = {}, the insanity seems to be solved.

**Comment by miraodasilva** [comment]
BTW I checked and the commit=False is only true on the last iteration, so this is working as intended.

**Comment by tssweeney** [comment]
Thank you for this very detailed breakdown - we appreciate your time helping to characterize the bug. It sounds like you have a workaround by moving the `logs` initialization call? Is that correct? We will continue to look into this and get back to you with any details.

**Comment by miraodasilva** [comment]
Yep, insanity in terms of logging thousands of file is fixed by moving that line. Which makes sense.

To reproduce this I would suggest trying to log the same audio and video file 100,000 times or something like that, and see if it gives off a similar bug.

No worries btw, thanks so much for the awesome tool, I think it's really great. And ofc thanks for your time.

**Comment by tssweeney** [comment]
Awesome, glad to hear you can continue with your work! Thanks for the kind feedback - we are always here to help. Please let us know if you run into further issues.
